crates/services/producer/src/block_producer.rs

```diff
@@ -268,26 +268,33 @@ where
     GasPriceProvider: GasPriceProviderConstraint,
     ConsensusProvider: ConsensusParametersProvider,
 {
-    // TODO: Support custom `block_time` for `dry_run`.
     /// Simulates multiple transactions without altering any state. Does not acquire the production lock.
     /// since it is basically a "read only" operation and shouldn't get in the way of normal
     /// production.
     pub async fn dry_run(
         &self,
         transactions: Vec<Transaction>,
         height: Option<BlockHeight>,
+        time: Option<Tai64>,
         utxo_validation: Option<bool>,
         gas_price: Option<u64>,
     ) -> anyhow::Result<Vec<TransactionExecutionStatus>> {
         let view = self.view_provider.latest_view()?;
-        let height = height.unwrap_or_else(|| {
-            view.latest_height()
-                .unwrap_or_default()
+        let latest_height = view.latest_height().unwrap_or_default();
+
+        let simulated_height = height.unwrap_or_else(|| {
+            latest_height
                 .succ()
                 .expect("It is impossible to overflow the current block height")
         });
 
-        let header = self._new_header(height, Tai64::now())?;
+        let simulated_time = time.unwrap_or_else(|| {
+            view.get_block(&latest_height)
```

Comment:
qq: should this use the latest height or the simulated height?

Reply:
Simulated height might not have a block, so defaulting to latest makes sense here to me. If the caller specifies a block height, and not a block time it is likely that the thing they are testing isn't using the block time. If someone needs to simulate a block with a specific height and timestamp, they can provide both.